ref(transactions): add functionality to sample transactions in save_event_transaction

JoshFerge opened the pull request:
first PR of https://github.com/getsentry/sentry/issues/81065

no behavior changes until option is turned on. need to change post_process as well before we can turn the option on. equivalent of https://github.com/getsentry/sentry/blob/b84d3d9238ab75822b076d16de1111c3035fe073/src/sentry/tasks/post_process.py#L614 in post_process

JoshFerge committed:
ref(transactions): add functionality to sample transactions in save_event_tx

lynnagara commented:
What are the implications of transactions sampling not happening exactly once?

Won't this potentially sample some messages in flight twice or not at all depending on the direction of option change

lynnagara commented:
I think https://github.com/getsentry/sentry/pull/81079 could probably be part of this same PR, otherwise all this option does is double sample.

JoshFerge commented:
> I think https://github.com/getsentry/sentry/pull/81079 could probably be part of this same PR, otherwise all this option does is double sample.

@lynnagara the goal of the PRs is to be reviewed separately, not to be deployed separately. i would never turn the option on without the other one deployed

JoshFerge commented:
>What are the implications of transactions sampling not happening exactly once?
> Won't this potentially sample some messages in flight twice or not at all depending on the direction of option change

from reading about the transactions sampler, i believe it is a feature that helps with parameterizing high cardinality transaction names. i don't believe double sampling (wouldn't happen as i've implemented)

or not having samples (would be the case for a few moments for this deploy) would have a large effect on the product.

https://develop.sentry.dev/api-server/application-domains/transaction-clustering/#accidental-erasure-of-non-identifiers

jjbayer commented on src/sentry/options/defaults.py:
```diff
@@ -2918,5 +2918,10 @@
+register(
+    "transactions.do_post_process_in_save",
+    type=Bool,
+    default=False,
```
Instead of using a boolean, should this be a rollout rate so we can do a gradual deterministic rollout based on project or org ID? See for example

https://github.com/getsentry/sentry/blob/303b9302ede7598e84832e063c84790f3d579318/src/sentry/grouping/strategies/message.py#L58-L60

jjbayer commented on src/sentry/event_manager.py:
```diff
@@ -2637,6 +2666,12 @@ def save_transaction_events(jobs: Sequence[Job], projects: ProjectsMapping) -> S
     with metrics.timer("save_transaction_events.send_occurrence_to_platform"):
         _send_occurrence_to_platform(jobs, projects)
 
+    with metrics.timer("save_transaction_events.sample_transactions"):
+        _sample_transactions_in_save(jobs, projects)
+
+    with metrics.timer("save_transaction_events.send_transaction_processed_signals"):
+        _send_transaction_processed_signals(jobs, projects)
```
Should we wrap this in a `try ... except` and log the exception? Previously `save_transaction_events` would not fail if any if this logic failed.

JoshFerge replied:
will wrap in try/catches. thanks

jjbayer reviewed (commented):
We originally put the "record transaction" logic in post process to move it off the critical path. But if that task is going to disappear completely, this makes sense.

armenzg commented on tests/sentry/event_manager/test_event_manager.py:
```diff
@@ -1540,6 +1541,133 @@ def test_transaction_event_span_grouping(self) -> None:
         assert spans == [{"hash": hash_values([span["description"]])} for span in data["spans"]]
 
+    @override_options({"transactions.do_post_process_in_save": True})
+    def test_transaction_sampler_and_recieve(self) -> None:
```
```suggestion
    def test_transaction_sampler_and_receive(self) -> None:
```

JoshFerge commented:
> We originally put the "record transaction" logic in post process to move it off the critical path. But if that task is going to disappear completely, this makes sense.

Yeah, that makes. Unfortunately at this point the load we're putting on RabbitMQ and potential nstability it causes isn't worth having a whole other task for transactions. I think once we have more durable execution with task workers, it's certainly worth investigating breaking up again.

armenzg commented on tests/sentry/event_manager/test_event_manager.py:
```diff
@@ -1540,6 +1541,133 @@ def test_transaction_event_span_grouping(self) -> None:
+    @override_options({"transactions.do_post_process_in_save": True})
+    @patch("sentry.signals.transaction_processed.send_robust")
+    @patch("sentry.ingest.transaction_clusterer.datasource.redis._record_sample")
+    def test_transaction_sampler_and_recieve_mock_called(
```
```suggestion
    def test_transaction_sampler_and_receive_mock_called(
```

armenzg commented on tests/sentry/event_manager/test_event_manager.py:
```diff
@@ -1540,6 +1541,133 @@ def test_transaction_event_span_grouping(self) -> None:
         # the basic strategy is to simply use the description
         assert spans == [{"hash": hash_values([span["description"]])} for span in data["spans"]]
 
+    @override_options({"transactions.do_post_process_in_save": True})
```
What value does this test add in contrast with the next one? I think they're the same but the next one tests for the two calls.

JoshFerge replied:
there is a comment below, but the next test ensures that the functions get called with the correct mocks, while this one ensures that the functions themselves don't error / run correctly. (didn't see a great way to validate that the side effects those functions call), so have two.

JoshFerge commented on src/sentry/event_manager.py:
```diff
@@ -2512,6 +2517,39 @@ def _detect_performance_problems(
+
+            project = job["event"].project
+            transaction_processed.send_robust(
+                sender=None,
```
i need to determine if setting a sender is important. i don't believe it is though.

markstory reviewed (commented):
Looks good with #81079 so that we aren't double sampling.

JoshFerge committed:
ref(transactions): option for not doing transactions sampling in post_process (#81079)

in lockstep with https://github.com/getsentry/sentry/pull/81077/, stops
doing work in post_process on transactions. part of
https://github.com/getsentry/sentry/issues/81065

JoshFerge commented:
> Looks good with #81079 so that we aren't double sampling.

merged that PR into this one so it's easier to read / can assure they go out together. (won't do anything until the option is modified)

JoshFerge commented:
rollout strategy: https://github.com/getsentry/sentry/issues/81065#issuecomment-2491767459

wedamija commented:
We also log https://github.com/getsentry/sentry/blob/d25bb2945c50ef50ccc5fbff57929e901b0b4fbf/src/sentry/tasks/post_process.py#L653-L659 as part of post process. I'm not sure if there are any datadog monitors or stat pages that will be affected by removing this. Might be worth adding this to the end of save event for transactions?

lynnagara commented:
We have to be very careful about how this impacts performance of save event. I think we should look into a couple of things first before making this change:

- Can we collect timings on transactions post processing is first before we move it into save event? I think that whether it is safe to do depends on the timing.
- Secondly, signals aren't allowed in the save event task, as it is too easy for additional work to be added there, and. Can we remove those?
- The volume on this pipeline is expected to increase up to 4 fold in the coming months and we would need to understand what's needed to scale if save event transactions job gets slower

wedamija commented:
> We have to be very careful about how this impacts performance of save event. I think we should look into a couple of things first before making this change:
> 
> * Can we collect timings on transactions post processing is first before we move it into save event? I think that whether it is safe to do depends on the timing. If it's above, say, 10 milliseconds, i think we should maybe consider some other options?
> * Secondly, signals aren't allowed in the save event task, as it is too easy for additional work to be added there, and. Can we remove those?
> * The volume on this pipeline is expected to increase up to 4 fold in the coming months and we would need to understand what's needed to scale if save event transactions job gets slower

I agree we should time it to be safe, but I think it's unlikely that it really takes very long compared to https://github.com/getsentry/sentry/blob/9ee399c5365a1e44d98b92f617249eb9bd220385/src/sentry/event_manager.py#L2635 which we do in in `save_transaction_events`. This does a lot of relatively expensive analysis on all the details of the transaction.

JoshFerge commented on src/sentry/event_manager.py:
```diff
@@ -2512,6 +2517,39 @@ def _detect_performance_problems(
+                continue
+
+            project = job["event"].project
+            transaction_processed.send_robust(
```
TODO: remove signal, call any receivers directly inline.

JoshFerge reviewed (commented):
TODO: let's use two options for the rollout

JoshFerge committed:
split into separate flags; don't use signal

JoshFerge commented:
we spent some time looking at the performance of the signals / transactions clustering on current telemetry in post_process, and both looked to be sub 10ms average, with no large outliers. this PR includes timing telemetry so we can monitor perf in S4S to confirm before any wider rollout takes place.

also confirmed that transaction clustering doesn't look at anything in clickhouse, it only uses redis.

i've split the option used to roll this out into two, so we can roll out by turning on `transactions.do_post_process_in_save`, confirming looks good (in this time we'll also do the same work in post_process, but doing things double here are fine)

after we confirm looks good, we can flip on `transactions.dont_do_transactions_logic_in_post_process`, which will stop doing the work in post_process.

lynnagara commented:
Currently there are 3 things that happen in post processing:
- transactions clustering
- signal + onboarding tasks
- delete from rc-processing

I think instead of splitting these under different options you can combine this under a single one `transactions.do_post_process_in_save` and only checking that option in save_event_transaction and not in post processing at all. I believe this works without any additional checks in post process because if you do everything in including deletion here then post process will never run anyway. The relevant line that prevents this from being re- post-processed is here: https://github.com/getsentry/sentry/blob/2e30c63b2e42fe2e6e11fa099bad827ade862668/src/sentry/tasks/post_process.py#L528.

I think there are a few benefits of this approach:
- Consistency. You do not have a minute or two when this option gets flipped where the events that are in the middle of the pipeline may skip this processing
- Less complexity
- All combinations of options are valid. If there is only one option then turning it up and down is harmless. It's easy for anyone (like SRE) to flip back in an incident. The other way requires more knowledge of the code to ensure an invalid combination isn't picked.

What do you think?

JoshFerge committed:
consolidate options, remove post_process logic, delete from rc-processing

JoshFerge commented:
> Currently there are 3 things that happen in post processing:
> 
> * transactions clustering
> * signal + onboarding tasks
> * delete from rc-processing
> 
> I think instead of splitting these under different options you can combine this under a single one `transactions.do_post_process_in_save` and only checking that option in save_event_transaction and not in post processing at all. I believe this works without any additional checks in post process because if you do everything in including deletion here then post process will never run anyway. The relevant line that prevents this from being re- post-processed is here:
> 
> https://github.com/getsentry/sentry/blob/2e30c63b2e42fe2e6e11fa099bad827ade862668/src/sentry/tasks/post_process.py#L528
> 
> .
> I think there are a few benefits of this approach:
> 
> * Consistency. You get exactly once processing without much work. You do not have a minute or two when this option gets flipped where the events that are in the middle of the pipeline may skip this processing
> * Less complexity
> * All combinations of options are valid. There is only one option and turning it up and down is harmless. It's easy for anyone (like SRE) to flip back in an incident. The other way requires more knowledge of the code to ensure an invalid combination isn't picked.
> 
> What do you think?

i've gone ahead and implemented this -- it's much cleaner. thank you for the suggestion. closing the follow up PR.

JoshFerge committed:
add assertion for case when flag is off

JoshFerge commented:
updated rollout plan: https://github.com/getsentry/sentry/issues/81065#issuecomment-2491767459

JoshFerge committed:
consolidate methods, use event_id instead of project_id for rollout

lynnagara commented on src/sentry/tasks/store.py:
```diff
@@ -582,6 +583,16 @@ def _do_save_event(
             raise
 
         finally:
+            if (
+                consumer_type == ConsumerType.Transactions
+                and event_id
+                and in_rollout_group("transactions.do_post_process_in_save", event_id)
+            ):
+                # we won't use the transaction data in post_process
+                # so we can delete it from the cache now.
+                if cache_key:
+                    processing_store.delete_by_key(cache_key)
+
```
just double checking - should this be in the `finally` block, or in an `else` block?

do we want to run it if if we hit the save event exception above?

JoshFerge replied:
i think we'd always want to do the processing store deletion -- otherwise we risk filling up the processing store in the case where we're not running transactions through post process

lynnagara reviewed (approved):
:shipit:
